cubix-tree-sitter/gen-ast/app/TreeSitter/GenerateAst/Internal/CodeGen.hs

```diff
@@ -0,0 +1,203 @@
+  parallel. This works for strictly specified ASTs, where each node has a known
+  number of children, but breaks down when nodes are allowed optional children,
+  children that contain lists of nodes, etc.
+-}
+
+-- TODO:
+--
+-- 1. Restructure parser to parse children as '[SomeNode]' and only then attempt to unpack
+--    that list to the appropriate instance of 'ChildList' so that any mismatch at this point
+--    can result in a 'VirtualError' as opposed to a 'SortMismatch'.
+--
+-- 2. Add `Extra` type to represent extra nodes; `Extra sort` wraps a `Node sort` and a proof that
+--    the `sort` is an "extra" sort. Permit occurances of `Extra` nodes before, after, and between
+--    all child nodes.
+--
+-- 3. Add callbacks for error and missing nodes. Distinguish between missing nodes and missing text.
+--    (The latter is not represented in the ast, but certainly constitutes a parsing failure.)
+--    Refine traversal functions, e.g., `gotoNextSibling`, to invoke these callbacks for all error
+--    and missing nodes. This requires changing `gotoNextSibling` to not call `gotoParent` until the
+--    end of the list of siblings is reached. The contract for these callbacks is that they are only
+--    called when the node has changes.
+--
+-- 4. Analyse the current tree traversal to see whether it is bottom-up or top-down, since the former
+--    is desirable with regards to memory usage.
+
+--------------------------------------------------------------------------------
+-- Template Parser and Renderer
+--------------------------------------------------------------------------------
+
+data Metadata = Metadata
+  { startRuleName :: RuleName
+  , moduleName :: Maybe Text
+  , pretty :: Bool
+  }
+
+generateAst :: Metadata -> Grammar -> FilePath -> Text -> Either String Text
+generateAst Metadata{..} grammar templateFile template = errorOrModule
+ where
+  defaultModuleName = snakeToCase Upper grammar.name <> "Ast"
+  metadataContext =
+    Context . M.fromList $
+      [ ("startRuleName", textToVal startRuleName)
+      , ("moduleName", textToVal $ fromMaybe defaultModuleName moduleName)
+      , ("pretty", BoolVal pretty)
+      ]
+  dataTypesContext =
+    Context . M.fromList $
+      [ ("dataTypes", toVal dataTypes)
+      , ("startSort", maybe NullVal (\(Data name _, _) -> toVal name) (uncons dataTypes))
+      ]
+  dataTypes = toDataTypes startRuleName grammar
+  context = metadataContext <> dataTypesContext
+  errorOrModule = renderTemplate templateFile template context
+
+data RenderState = InText | InTemplate [Text]
+
+renderTemplate :: (ToContext Text b) => FilePath -> Text -> b -> Either String Text
+renderTemplate templateFile templateText context =
+  fmap (TL.toStrict . TLB.toLazyText) . go InText . zip [1 ..] . T.lines $ templateText
+ where
+  go :: RenderState -> [(Int, Text)] -> Either String TLB.Builder
+  go InText [] = pure mempty
+  go InText ((_n, ln) : lns)
+    | ln == "{-+}" = go (InTemplate mempty) lns
```

Comment:
I must say, I'm very confused by these templates. Did you implement the template expansion system yourself, or is it a library? It looks like the former, but I'm seeing "{for" notation in there that's not implemented here.